Add new `CASE` expression.

ndm2 opened the pull request:
Since the previous discussions regarding a new syntax didn't really go anywhere, I thought it might be best to just throw something new into the room and go from there. This is more or less directly ported from some private code of mine, for now I've just made some minor adjustments and added some specific test for it to better fit in the core, so please don't hesitate to trash it, pick only what you like, or even outright completely dismiss it - everything is for sale!

I know that especially the previously suggested `->when()->then()` syntax was controversial in regard to defining multiple `WHEN THEN`'s, but from my experience I can say that it's quite nice to use in most situations, and still very much readable if you don't have looots of `WHEN THEN`'s.

The expressions make use of the current query's type map and automatically bind conditions accordingly. Types for non-condition values are being inferred when possible, and bound accordingly too. I bring that up because the second argument of the `when()`, `then()`, `else()` methods is reserved for specifying types, which is another reason why I preferred splitting `WHEN` and `THEN` into multiple methods, as otherwise the syntax would again become a bit cumbersome when it's something like `add($when, $whenType, $then, $thenType)`.

So that being said, in its current state there's support for a call-order based syntax:

```php
$queryExpression
    ->case()
    ->when(['Table.column' => true])
    ->then('Yes')
    ->when(['Table.column' => false])
    ->then('No')
    ->else('Maybe');
```

(which will scream at you when used incorrectly, for example when doing something like `when(1)->else(2)` or `then(1)->then(2)`), and a callable syntax with self-contained `WHEN ... THEN ...` expressions:

```php
$queryExpression
    ->case()
    ->when(function (WhenThenExpressionInterface $whenThen) {
        return $whenThen
            ->when(['Table.column' => true])
            ->then('Yes');
    })
    ->when(function (WhenThenExpressionInterface $whenThen) {
        return $whenThen
            ->when(['Table.column' => false])
            ->then('No');
    })
    ->else('Possibly');
```

Additionally to the searched case variant, the simple case variant (eg `CASE value WHEN ...`) is supported too:

```php
$queryExpression
    ->case()
    ->value($query->identifier('Table.column'))
    ->when(true)
    ->then('Yes')
    ->when(false)
    ->then('No')
    ->else("Yesn't");
```

And to sprinkle in some more potential for contention, column names _must_ be passed as `IdentifierExpression` instances, as strings will simply be bound as strings.

refs #11793, #14753, #11896

ndm2 committed:
Add new `CASE` expression.

othercorey commented:
I see why you have `else()` and `getElse()` so it reads like the other sql fluent api. However, it feels like an awkward choice and not something we do elsewhere.

markstory commented:
> I see why you have else() and getElse() so it reads like the other sql fluent api. However, it feels like an awkward choice and not something we do elsewhere.

I think it is more consistent with the rest of the ORM though, and that is more important. We aimed to keep the `QueryBuilder` interface as close to SQL keywords as we can.

othercorey commented:
>  We aimed to keep the `QueryBuilder` interface as close to SQL keywords as we can.

Yea I wouldn’t change else(). I was thinking if we can combine the getters.

markstory commented on src/Database/Expression/CaseExpressionInterface.php:
```diff
@@ -0,0 +1,332 @@
+     *
+     * @return \Cake\Database\Expression\WhenThenExpressionInterface[]
+     */
+    public function getWhen(): array;
```
Do you need all the getter methods? Could something like `Query::clause()` work as a general purpose accessor.

ndm2 replied:
They're not necessarily needed, no. I don't remember the exact use case that made me add them initially, that was quite some time ago. The type getters were introduced later, actually _needed_ was just `WhenThenExpression::getThenType()` so that the return type for `CaseStatementExpression` could be guessed, and then this lonely getter seemed weird, so I added some friends.

Something like `Query::clause()` could definitely work. Not sure what it could look like for types, maybe the types would be separate "clauses", or a clause would be an associative array that holds the value and the type... 🤔

markstory replied:
Couldn't the clauses of this be `when`, `then` and `else`?

ndm2 replied:
Sure, I'm just unsure about how to obtain the accompanying types.

ndm2 replied:
I don't really have a use case for any of the types other than the type for the nested `WhenThenExpression::then()`.

How about not exposing any types other than that one, it could be a method like `WhenThenExpression::getResultType()`, making things a bit more abstract/generic?

ndm2 replied:
So I've tried to find the project that was the initial reason for me to add the expression part getters, but I can't find it, and unfortunately the commit for the expression code only says "_Add expression part getters_" - that's what you get for being lazy :)

So just let me add some kind of clause method, and the more generic result type getter, and see how that is received when it's actually tangible.

ndm2 replied:
Testing type inference without the getters is not going to work, I thought I'd test the bindings, but some values are not bound while there's still an inferred type, and testing them indirectly is cumbersome and error prone. I'll keep it on the backburner, let's first evaluate the clause getter.

markstory replied:
> How about not exposing any types other than that one, it could be a method like WhenThenExpression::getResultType(), making things a bit more abstract/generic?

Limiting the scope of the methods we expose sounds good to me.

ndm2 replied:
OK, I just need to figure a way to test the type inference properly without all those type getters.

ndm2 replied:
I've settled for stubs 🤷‍♂️ So now there's only one type getter, which is actually required and made use of.

markstory commented on src/Database/Expression/CaseExpressionInterface.php:
```diff
@@ -0,0 +1,332 @@
+     *
+     * Another way to safely pass user data is when using a conditions
+     * array, and passing user data only on the value side of the array
+     * entries, which will cause them to be bound:
```
:clap: yay

markstory commented on src/Database/Expression/CaseExpressionInterface.php:
````diff
@@ -0,0 +1,332 @@
+     * ```
+     * $case
+     *     ->then('Yes')
+     *     // ...
````
Isn't this scenario invalid as then() can only be called after a when()?

ndm2 replied:
It is, all the examples here are ment to show invalid usage.

markstory commented on src/Database/Expression/CaseStatementExpression.php:
```diff
@@ -0,0 +1,409 @@
+
+        $this->_whenBuffer = null;
+
+        $this->when($whenThen);
```
Couldn't you skip some complexity by appending to `_when` here?

ndm2 replied:
Yeah, the method call doesn't seem to be needed, probably some refactoring leftover.

markstory commented on src/Database/Expression/CaseStatementExpression.php:
```diff
@@ -0,0 +1,409 @@
+    /**
+     * @inheritDoc
+     */
+    public function getElseType(): ?string
```
This could be part of the `clause()` method too.

markstory commented on src/Database/Expression/CaseStatementExpression.php:
```diff
@@ -0,0 +1,409 @@
+    /**
+     * @inheritDoc
+     */
+    public function getElse()
```
This could be part of the clause() method.

markstory commented on src/Database/Expression/CaseStatementExpression.php:
```diff
@@ -0,0 +1,409 @@
+                    'Cannot compile incomplete `\%s` expression, there are no `WHEN ... THEN ...` statements.',
+                    CaseExpressionInterface::class
+                )
+            );
```
Nice work with the thorough error handling here :clap:

markstory commented on src/Database/Expression/CaseExpressionTrait.php:
```diff
@@ -0,0 +1,96 @@
+use Cake\Database\ValueBinder;
+use DateTimeInterface;
+
+trait CaseExpressionTrait
```
Can we mark this trait as internal? This doesn't feel like the kind of trait that userland code should be depending on.

ndm2 replied:
Absolutely!

markstory commented on src/Database/Expression/WhenThenExpression.php:
```diff
@@ -0,0 +1,321 @@
+    /**
+     * @inheritDoc
+     */
+    public function getThen()
```
Could be another `clause()` method on this class too.

markstory commented on src/Database/Expression/WhenThenExpression.php:
```diff
@@ -0,0 +1,321 @@
+     *
+     * @return \Cake\Database\TypeMap
+     */
+    protected function getTypeMap(): TypeMap
```
Do you need a method to access the property? Couldn't you use the property directly?

ndm2 replied:
Not necessarily needed, no. It is to satisfy `CaseExpressionTrait::inferType()`, which uses `getTypeMap()` since it's already available from the `TypeMapTrait`, which in turn was just used out of lazyness, and resulted in no type map being set by default for `CaseStatementExpression`, hence making me use `getTypeMap()` in the first place which creates a map if none is present 😆

The `CaseStatementExpression` constructor could certainly create a default type map, then `CaseExpressionTrait::inferType()` could use the property.

markstory commented on src/Database/Expression/WhenThenExpression.php:
```diff
@@ -0,0 +1,321 @@
+     *
+     * @var \Cake\Database\TypeMap
+     */
+    protected $_typeMap;
```
While most of the core uses `_` prefixed protected variables, you don't have to use the `_` for new code if you don't want to.

ndm2 replied:
Totally forgot about the new convention.

markstory reviewed (commented):
I haven't gotten through the whole thing, but this is my thoughts on the first 6 of 13 files.

ndm2 committed:
Mark case expression trait as internal.

ndm2 committed:
Drop the old underscore prefix convention.

ndm2 committed:
Refactor to access the type map via the property.

ndm2 committed:
Store expression directly instead of unnecessarily calling `when()`.

ndm2 committed:
Consolidate the expression part getters into a single method.

markstory reviewed (commented):
Looks good to me. If you want I can write some docs for this.

ADmad commented on src/Database/Expression/CaseStatementExpression.php:
```diff
@@ -0,0 +1,398 @@
+    /**
+     * The else part result value.
+     *
+     * @var mixed|null
```
```suggestion
     * @var mixed
```

ndm2 replied:
Not sure why this is still using mixed in the first place, as the accepted values are known.

ADmad commented on src/Database/Expression/CaseStatementExpression.php:
```diff
@@ -0,0 +1,398 @@
+    /**
+     * The `WHEN ... THEN ...` expressions.
+     *
+     * @var \Cake\Database\Expression\WhenThenExpressionInterface[]
```
```suggestion
     * @var array<\Cake\Database\Expression\WhenThenExpressionInterface>
```

ndm2 commented:
> Looks good to me. If you want I can write some docs for this.

Please be my guest, just take into account that there's still some small change coming with regards to the type getters.

ndm2 committed:
Fix property docblock type hints.

ndm2 committed:
Remove type getters.

Only keep a getter for the result type of nested `WhenThenExpression`'s,
it's somewhat internal and required to guess the `CASE` expression's
return type.

ndm2 committed:
Fix CS.

markstory commented:
> Please be my guest, just take into account that there's still some small change coming with regards to the type getters.

Do those need to be in the book? Their primary purpose seems to be for testing.

ndm2 commented:
No no, they don't, I meant that maybe, possibly, the removal of the getters could've somehow warranted changes to the existing API... but it turned out to not be the case.

othercorey commented on src/Database/Expression/CaseStatementExpression.php:
```diff
@@ -0,0 +1,382 @@
+            $type = $this->inferType($result);
+        }
+
+        $this->whenBuffer = null;
```
When is this needed? It throws an exception at the start of function.

ndm2 replied:
Scratch my previous comment, we're in `else()` here, not in `then()`, sorry. Yeah, this isn't needed, looks like copy pasta leftover.

othercorey commented on src/Database/Expression/CaseStatementExpression.php:
```diff
@@ -0,0 +1,382 @@
+        $this->whenBuffer = null;
+
+        $this->else = $result;
+        $this->elseType = $type;
```
This allows overwriting else, should we warn?

ndm2 replied:
That would be a precedent, wouldn't it? I can't really think of a use case that would _require_ overwriting, but no other expression currently prohibits this AFAICT, so I'm unsure.

othercorey commented on src/Database/Expression/CaseStatementExpression.php:
```diff
@@ -0,0 +1,382 @@
+     */
+    public function when($when, $type = null)
+    {
+        if ($this->whenBuffer !== null) {
```
Should we enforce the order expected in sql and error on `else()->when()` ?

ndm2 replied:
I guess it could be argued that it would make the whole thing more consistent as far as the API's behavior goes, but it might be a bit overkill, as we'd move away from an actual technical need. I'm not necessarily opposed to it though.

ndm2 committed:
Remove unnecessary emptying of the buffer property.

othercorey commented:
Looks good to me.

othercorey commented on src/Database/Expression/QueryExpression.php:
```diff
@@ -342,9 +342,21 @@ public function in($field, $values, $type = null)
+     *
+     * @return \Cake\Database\Expression\CaseExpressionInterface
+     */
+    public function case(): CaseExpressionInterface
```
Is there a reason the `value()` parameter isn't supported here?

ndm2 replied:
Not really... I just grew accustom to the `value()` method (which once replaced a really awkward constructor syntax), and when I ported this over to the core, I just didn't think of it 🤷‍♂️

Would you like me to add support for it?

othercorey replied:
I think it would read more like sql with it unless you have technical limits here.

ndm2 replied:
Thinking about it... `null` would technically be a problem here, as we cannot distinguish between no value passed and explicit `null` passed. Even though SQL `NULL` cannot be compared against (`WHEN NULL` will never match), it is actually a valid case value.

People would need to be aware of that behavior, and use the `value()` method in case they dynamically pass a value that could be `null`.

othercorey replied:
When would you need to switch on null?

ndm2 replied:
I'm not sure that you ever actually _need_ to, but I also don't really know why it's valid on SQL level in the first place, maybe just because failing hard when a column or function return value is `NULL` would not be nice, or maybe there is actually use case for that behavior.

On PHP level it could easily happen with dynamic values. We could forbid `null` for the value, not sure which behavior people would prefer.

othercorey replied:
You can use func_num_args() if NULL support is desired.

ndm2 replied:
Oh boy, there's a function that I haven't used in 20 years, I pretty much forgot it even existed. Well OK, then I don't think I see any technical hurdles.

ndm2 committed:
Add support for setting the `CASE` value via the factory method.

othercorey commented:
Looks good to me.

markstory commented on tests/test_app/TestApp/Database/Expression/CustomWhenThenExpression.php:
```diff
@@ -0,0 +1,23 @@
+
+use Cake\Database\Expression\WhenThenExpression;
+
+class CustomWhenThenExpression extends WhenThenExpression
```
I hope no one needs to do this but its good they can.

ADmad commented on tests/TestCase/Database/QueryTest.php:
```diff
@@ -1872,16 +1872,18 @@ public function testSelectOrderAsc(): void
-                );
-            })
-            ->orderAsc('id');
+        $this->deprecated(function () use ($query) {
```
Would be nice if the tests where updated to use the new method instead. Can add a separate test method to check for deprecated usage with `@deprecated` tag if needed. This would make cleaning up the deprecated code a lot easier after merging into `5.x` branch.

ndm2 replied:
Since there's integration tests in `\Cake\Test\TestCase\Database\QueryTests\CaseExpressionQueryTest` (I think they cover all `addCase()` ones except for subquery usage), would it suffice to just mark the tests here with `@deprecated`?

ADmad replied:
Yes, if these are specifically for checking `CASE` clause usage and they are already covered in `CaseExpressionQueryTest` then we can just mark these methods as `@deprecated` and drop them in `5.x`.

ndm2 replied:
Ok, some of it actually had to be updated, as it wasn't specifically testing `addCase()`, but rather expression usage.
